torch/csrc/jit/passes/frozen_concat_linear.cpp

```diff
@@ -0,0 +1,186 @@
+      Value* op_value = n->inputs().at(0);
+      auto weight = n->namedInput("weight");
+      auto bias = n->namedInput("bias");
+
+      if (!constant_as<Tensor>(weight).has_value() ||
+          !constant_as<Tensor>(bias).has_value()) {
+        continue;
+      }
+      if (weight->type() == NoneType::get() ||
+          bias->type() == NoneType::get()) {
+        continue;
+      }
+
+      if (tensorUsers.find(op_value) == tensorUsers.cend()) {
+        tensorUsers.insert({op_value, std::list<Node*>()});
+      }
+      tensorUsers.find(op_value)->second.push_back(n);
+    }
+
+    // Check the uses of a tensor to find ones that can be combined
+    for (auto& kv : tensorUsers) {
+      Value* op_value = kv.first;
+      auto& uses = kv.second;
+
+      while (!uses.empty()) {
+        auto it_first = uses.begin();
+        Node* base_node = *it_first;
+        uses.erase(it_first);
+
+        auto weight_type = constant_as<Tensor>(base_node->namedInput("weight"))->dtype();
+        auto bias_type = constant_as<Tensor>(base_node->namedInput("bias"))->dtype();
+        std::vector<Node*> matching_nodes;
+        matching_nodes.push_back(base_node);
+
+        // Now iterate over the rest of the users of the set to
+        // see if there is anything that we can coaleasce with.
+        for (auto it = uses.begin(); it != uses.end();) {
+          auto candidate_node = *it;
+          auto candidate_weight = constant_as<Tensor>(candidate_node->namedInput("weight"));
+          auto candidate_weight_type = candidate_weight->dtype();
+          auto candidate_bias = constant_as<Tensor>(candidate_node->namedInput("bias"));
+          auto candidate_bias_type = candidate_bias->dtype();
+
+          // For now we will just keep it simple and require matching types
+          // Type promotion mgiht cause performance to actually decrease.
+          if (weight_type != candidate_weight_type || bias_type != candidate_bias_type) {          
+            it++;
+            continue;
+          }
+
+          bool can_move_before_all = true;
+          for (auto n : matching_nodes) {
+            can_move_before_all &=
+                getAliasDb()->couldMoveBeforeTopologically(candidate_node, n);
+          }
+          if (!can_move_before_all) {
+            it++;
+            continue;
+          }
+
+          // Found a node that is eligible for combination
+          auto prev = it;
+          it++;
+          uses.erase(prev);
+          matching_nodes.push_back(candidate_node);
+        }
+        if (matching_nodes.size() == 1) {
+          // Found no other compatible linear layers, do nothing.
+          continue;
+        }
+        graph_modified = true;
+
+        for (auto n : matching_nodes) {
+          if (n == base_node) {
+            continue;
+          }
+          getAliasDb()->moveBeforeTopologicallyValid(base_node, n);
+        }
+
+        // Now actually merge the candidate nodes
+        graph_->setInsertPoint(base_node);
```

Comment:
this API isnt met to be used because it modifies global state, use WithInsertPoint instead

Reply:
I didn't see a good way of changing the insert point (because I set this twice) without creating new scopes. 
I could heap allocate and delete a WithInsertPoint object, but that feels weird. 
Putting it in an anonymous scope makes it hard to use the other variables I make.